src/core/search/rax_tree.h

```diff
@@ -24,30 +24,37 @@ template <typename V> struct RaxTreeMap {
 
   // Simple seeking iterator
   struct SeekIterator {
-    friend struct FindIterator;
-
     SeekIterator() {
-      raxStart(&it_, nullptr);
-      it_.node = nullptr;
-    }
-
-    ~SeekIterator() {
-      raxStop(&it_);
+      it_.rt = nullptr;
     }
 
-    SeekIterator(SeekIterator&&) = delete;       // self-referential
-    SeekIterator(const SeekIterator&) = delete;  // self-referential
-
     SeekIterator(rax* tree, const char* op, std::string_view key) {
       raxStart(&it_, tree);
-      raxSeek(&it_, op, to_key_ptr(key), key.size());
-      operator++();
+      if (raxSeek(&it_, op, to_key_ptr(key), key.size())) {  // Successfuly seeked
+        operator++();
+      } else {
+        InvalidateIterator();
+      }
     }
 
     explicit SeekIterator(rax* tree) : SeekIterator(tree, "^", std::string_view{nullptr, 0}) {
     }
 
+    /* Remove copy/move constructors to avoid double iterator invalidation */
+    SeekIterator(SeekIterator&&) = delete;
+    SeekIterator(const SeekIterator&) = delete;
+    SeekIterator& operator=(SeekIterator&&) = delete;
+    SeekIterator& operator=(const SeekIterator&) = delete;
+
+    ~SeekIterator() {
+      if (IsValid()) {
+        InvalidateIterator();
+      }
+    }
+
     bool operator==(const SeekIterator& rhs) const {
+      if (!IsValid() || !rhs.IsValid())
```

Comment:
why do you prefer this notation?
you do not find `it_.rt == other.rt &&  it_.node == rhs.it_.node;` more readable?

Reply:
But this is not true. If both iterators are invalid, `it_.rt == rhs.it_.rt,` but `it_.node == rhs.it_.node` does not always hold true. We rely on this behavior when comparing a `FindIterator` with `RaxTreeMap::end()`, where `it_.node` is not set. And also we don't know where raxFind was stopped. In other words, we don't know the node that `it_.node` points to after `raxFind`.

Reply:
So it should be `(!it_.rt && !other.rt) || (it_.rt && other.rt && it_.node == rhs.it_.node)` . I don't think it is more readable